File: examples/PHP/ppqueue.php
<?php
/* 
 * Paranoid Pirate queue
 * 
 * @author Ian Barber <ian(dot)barber(at)gmail(dot)com>
 */
include "zmsg.php";

define("MAX_WORKERS", 100);
define("HEARTBEAT_LIVENESS", 3); //  3-5 is reasonable
define("HEARTBEAT_INTERVAL", 1); //  secs


class Queue_T implements Iterator{
	private $queue = array();
	
	/* Iterator functions */
	public function rewind() { return reset($this->queue); }
	public function valid() { return current($this->queue); }
	public function key() { return key($this->queue); }
	public function next() { return next($this->queue); }
	public function current() { return current($this->queue); }

	/*  
	 * Insert worker at end of queue, reset expiry
	 * Worker must not already be in queue
	 */
	public function s_worker_append($identity) {
		if(isset($this->queue[$identity])) {
			printf ("E: duplicate worker identity %s", $identity);
		} else {
			$this->queue[$identity] = microtime(true) + HEARTBEAT_INTERVAL * HEARTBEAT_LIVENESS;
		}
	}

	/*
	 * Remove worker from queue, if present
	 */
	public function s_worker_delete($identity) {
		unset($this->queue[$identity]);
	}

	/*
	 * Reset worker expiry, worker must be present
	 */
	function s_worker_refresh($identity) {
		if(!isset($this->queue[$identity])) {
			printf ("E: worker %s not ready\n", $identity);
		} else {
			$this->queue[$identity] = microtime(true) + HEARTBEAT_INTERVAL * HEARTBEAT_LIVENESS;
		}
	}
	
	/*
	 * Pop next available worker off queue, return identity
	 */
	public function s_worker_dequeue() {
		reset($this->queue);
		$identity = key($this->queue);
		unset($this->queue[$identity]);
		return $identity;
	}

	/* 
	 * Look for & kill expired workers
	 */
	public function s_queue_purge() {
		foreach($this->queue as $id => $expiry) {
			if(microtime(true) > $expiry) {
				unset($this->queue[$id]);
			}
		}
	}
	
	/*
	 * Return the size of the queue
	 */
	public function size() {
		return count($this->queue);
	}
}

//  Prepare our context and sockets
$context = new ZMQContext();
$frontend = new ZMQSocket($context, ZMQ::SOCKET_ROUTER);
$backend = new ZMQSocket($context, ZMQ::SOCKET_ROUTER);
$frontend->bind("tcp://*:5555");    //  For clients
$backend->bind("tcp://*:5556");    //  For workers
$read = $write = array();

//  Queue of available workers
$queue = new Queue_T();

//  Send out heartbeats at regular intervals
$heartbeat_at = microtime(true) +  HEARTBEAT_INTERVAL;

while(true) {
	$poll = new ZMQPoll();
	$poll->add($backend, ZMQ::POLL_IN);
	
	//  Poll frontend only if we have available workers
	if($queue->size()) {
		$poll->add($frontend, ZMQ::POLL_IN);
	}
	
	$events = $poll->poll($read, $write, HEARTBEAT_INTERVAL * 1000 ); // milliseconds
	
	if($events > 0) {
		foreach($read as $socket) {
			$zmsg = new Zmsg($socket);
			$zmsg->recv();
			
			//  Handle worker activity on backend
			if($socket === $backend) {
				$identity = $zmsg->unwrap();
				
				//  Return reply to client if it's not a control message
				if($zmsg->parts() == 1) {
					if($zmsg->address() == "READY") {
						$queue->s_worker_delete($identity);
						$queue->s_worker_append($identity);
 					} else if($zmsg->address() == 'HEARTBEAT') {
						$queue->s_worker_refresh($identity);
					} else {
						printf ("E: invalid message from %s%s%s", $identity, PHP_EOL, $zmsg->__toString());
					}
				} else {
					$zmsg->set_socket($frontend)->send();
					$queue->s_worker_append($identity);
				}
			} else {
				//  Now get next client request, route to next worker
				$identity = $queue->s_worker_dequeue();
				$zmsg->wrap($identity);
				$zmsg->set_socket($backend)->send();
			}
		}		
		
		if(microtime(true) > $heartbeat_at) {
			foreach($queue as $id => $expiry) {
				$zmsg = new Zmsg($backend);
				$zmsg->body_set("HEARTBEAT");
				$zmsg->wrap($identity, NULL);
				$zmsg->send();
			}
			$heartbeat_at = microtime(true) + HEARTBEAT_INTERVAL;
		}
		$queue->s_queue_purge();
	}
}

File: examples/Python/tornado_ioloop/ppqueue.py
#!/usr/bin/env python

"""
synopsis:
    Paranoid Pirate queue
    Original author: Daniel Lundin <dln(at)eintr(dot)org>
    Modified for async/ioloop: Dave Kuhlman <dkuhlman(at)davekuhlman(dot)org>
usage:
    python ppqueue.py
"""

import sys
from collections import OrderedDict
import time
from functools import partial
import zmq
from zmq.eventloop.future import Context, Poller
from zmq.eventloop.ioloop import IOLoop
from tornado import gen

HEARTBEAT_LIVENESS = 3     # 3..5 is reasonable
HEARTBEAT_INTERVAL = 1.0   # Seconds

#  Paranoid Pirate Protocol constants
PPP_READY = b"\x01"      # Signals worker is ready
PPP_HEARTBEAT = b"\x02"  # Signals worker heartbeat
FRONT_END_ADDRESS = 'tcp://*:5555'
BACK_END_ADDRESS = 'tcp://*:5556'


class Worker(object):
    def __init__(self, address):
        self.address = address
        self.expiry = time.time() + HEARTBEAT_INTERVAL * HEARTBEAT_LIVENESS


class WorkerQueue(object):
    def __init__(self):
        self.queue = OrderedDict()

    def ready(self, worker):
        self.queue.pop(worker.address, None)
        self.queue[worker.address] = worker

    def purge(self):
        """Look for & kill expired workers."""
        t = time.time()
        expired = []
        for address, worker in self.queue.items():
            if t > worker.expiry:  # Worker expired
                expired.append(address)
        for address in expired:
            print("W: Idle worker expired: %s" % address)
            self.queue.pop(address, None)

    def __next__(self):
        address, worker = self.queue.popitem(False)
        return address


@gen.coroutine
def run_queue(context):
    frontend = context.socket(zmq.ROUTER)    # ROUTER
    backend = context.socket(zmq.ROUTER)     # ROUTER
    frontend.bind(FRONT_END_ADDRESS)    # For clients
    backend.bind(BACK_END_ADDRESS)      # For workers
    poll_workers = Poller()
    poll_workers.register(backend, zmq.POLLIN)
    poll_both = Poller()
    poll_both.register(frontend, zmq.POLLIN)
    poll_both.register(backend, zmq.POLLIN)
    workers = WorkerQueue()
    heartbeat_at = time.time() + HEARTBEAT_INTERVAL
    while True:
        if len(workers.queue) > 0:
            poller = poll_both
        else:
            poller = poll_workers
        socks = yield poller.poll(HEARTBEAT_INTERVAL * 1000)
        socks = dict(socks)
        # Handle worker activity on backend
        if socks.get(backend) == zmq.POLLIN:
            # Use worker address for LRU routing
            frames = yield backend.recv_multipart()
            if not frames:
                break
            address = frames[0]
            workers.ready(Worker(address))
            # Validate control message, or return reply to client
            msg = frames[1:]
            if len(msg) == 1:
                if msg[0] not in (PPP_READY, PPP_HEARTBEAT):
                    print("E: Invalid message from worker: %s" % msg)
            else:
                yield frontend.send_multipart(msg)
            # Send heartbeats to idle workers if it's time
            if time.time() >= heartbeat_at:
                for worker in workers.queue:
                    msg = [worker, PPP_HEARTBEAT]
                    yield backend.send_multipart(msg)
                heartbeat_at = time.time() + HEARTBEAT_INTERVAL
        if socks.get(frontend) == zmq.POLLIN:
            frames = yield frontend.recv_multipart()
            if not frames:
                break
            frames.insert(0, next(workers))
            backend.send_multipart(frames)
        workers.purge()


@gen.coroutine
def run(loop):
    context = Context()
    while True:
        yield run_queue(context)


def main():
    args = sys.argv[1:]
    if len(args) != 0:
        sys.exit(__doc__)
    try:
        loop = IOLoop.current()
        loop.run_sync(partial(run, loop))
    except KeyboardInterrupt:
        print('\nFinished (interrupted)')


if __name__ == '__main__':
    main()
